block: fix FilterGUID

hugelgupf opened the pull request:
Hey @insomniacslk @tfg13 @pmazzini do y'all use this logic? Because it is broken as far as I can tell. Take a look at this small change emphasizing what this actually does.

Since the only usage of this doesn't do what the code seems to think it does -- can I get rid of this function and fix it so it returns the partition?

mpictor commented:
I do have a need for filtering partitions by GUID in #1704. Looks like `storage.PartitionsByGUID` became `block.BlockDevices.FilterGUID` but didn't actually live up to its name?

hugelgupf commented:
Yeah, never did. Let's implement a real one?

insomniacslk commented:
I'll look at this shortly, thanks for letting me know :)

insomniacslk commented:
@hugelgupf `FilterESP` is not used on our side, I implemented it with the intention to use it at some point but as far as I'm concerned it can be removed. Could you please also remove the comment about `FilterESP` in `cmds/boot/localboot/main.go` ?

hugelgupf commented:
@insomniacslk my point was more that mountByGuid in localboot does not work as written. That means path mode doesn't work either. Or does it?

insomniacslk commented:
On memory it should work, let me check on a laptop

insomniacslk commented:
I'm not sure what's the mismatch between the function name/comment and its behaviour - it's looking for (the first) partition with a given GUID, mounts it, and returns the mount point as expressed in the function comment. What is the behaviour you expect?

hugelgupf commented:
If you read the code of FilterGUID, and the test I wrote, you'll see it does not find a partition. It finds a DISK.

FilterGUID(ESP) = /dev/sda, NOT /dev/sda1

As written.

hugelgupf commented:
What you describe is the behavior I expect, it's *not the behavior implemented*.

tfg13 commented:
It seems this was already wrong in the very first commit 2 years ago: https://github.com/u-root/u-root/commit/d74b17e8fedc4641f148bf7860f46809f6aba2ff, `FilterEFISystemPartitions` also returns the device instead of the partition, I suppose it went unnoticed because the function was essentially unused: https://github.com/u-root/u-root/commit/d74b17e8fedc4641f148bf7860f46809f6aba2ff#diff-b360ede5ae0435be5b7f7ce29d5ceb65R47

Not really sure how this went unnoticed for so long. I think we are not supplying a specific GUID, so it searches all partitions for grub configs, which then works fine and it never uses this function.

hugelgupf commented:
Does that mean we can remove the path mode or do you want it fixed?

We should have a fixed FilterGUID function definitely, I'm just curious about what you want from localboot given this discovery.

tfg13 commented:
Should we just fix the function (to return the partition) and change the name back? From looking at `mountByGUID`, it does not look like it would work right now, or does it?

insomniacslk commented:
> If you read the code of FilterGUID, and the test I wrote, you'll see it does not find a partition. It finds a DISK.
> 
> FilterGUID(ESP) = /dev/sda, NOT /dev/sda1
> 
> As written.

Sorry, I didn't catch what you meant - yes, it finds a disk, not a partition, and that's not the right behaviour. If someone else needs that, it should be fixed, otherwise we can remove it now and reintroduce it later

tfg13 commented:
I feel path mode might just work if we fix the function to actually return the partition

hugelgupf commented:
It'll work, but it's obviously not used, so what's the gain in fixing it rather than deleting it? 

(I wanna fix it & delete code, unless you want to start using path mode some time)

insomniacslk commented:
We are not going to use it soon, so this is OK to remove for me

hugelgupf commented:
I fixed it, and I deleted localboot code. If you want to keep it, I'm fine with that, too -- just a proposal. PTAL

mpictor commented on pkg/mount/block/blockdev.go:
```diff
@@ -387,32 +386,42 @@ func (b BlockDevices) FilterZeroSize() BlockDevices {
-
-// FilterGUID returns a list of BlockDev objects whose underlying
-// block device has the given GPT partition GUID.
+// FilterGUID returns a list of partition BlockDev objects whose partition GUID
```
I'd reword a bit:
```diff
-// FilterGUID returns a list of partition BlockDev objects whose partition GUID
+// FilterGUID returns a list of partition BlockDev objects whose partition type GUID
```

`part.Type` on line 402 is the partition's type guid, not its unique partition guid.

mpictor replied:
I propose renaming this function, perhaps to `FilterPartType()` and creating a new one `FilterPartID()` which compares `part.Id` rather than `part.Type`.

hugelgupf replied:
good point. changing it.

hugelgupf replied:
done!

hugelgupf commented:
> I fixed it, and I deleted localboot code. If you want to keep it, I'm fine with that, too -- just a proposal. PTAL

@insomniacslk @tfg13 i'd feel more comfortable merging this with one of your stamps of approval :)

insomniacslk commented:
Right now this is safe to merge, I'm discussing internally when we will need path mode again (the "if we need it" being almost certainly "yes"), and will get back to this

insomniacslk commented:
so - I'm reconsidering this a bit (and sorry for not verifying it earlier). Systemboot currently allows booting from grub config or from VPD variables (see [pkg/systembooter.GetBootEntries](https://github.com/u-root/u-root/blob/master/pkg/boot/systembooter/bootentry.go#L61) ). While removing these flags in localboot would not break that, it is still useful to have for manual troubleshooting (which we didn't need in a long time now).

Regarding the "it could have never worked", I'm baffled - I'm quite sure I've used this on my test devices, and I'm wondering if I messed up with the boot image (i.e. writing it to /dev/sda instead of /dev/sda1).

In conclusion, it is safe to merge this now, but I may need this feature in the future. If this is an easy fix here, I'd leave it, if not let's merge this, and I'll reintroduce it later when necessary. How does that sound?

hugelgupf commented:
I'll revert the deletion. It should fix the localboot case.

hugelgupf commented:
@mpictor PTAL -- now with FilterPartID :)

insomniacslk commented:
Thank you Chris!
